### PR TITLE
infra: refine fuzztest work

### PR DESCRIPTION
**Pull request by DavidKorczynski**
Enable Fuzztest fuzzers for Tensorflow.

This depends on https://github.com/google/fuzztest/pull/79 and eventually a PR on the tensorflow repo with the (to be refined) diff in this PR.

Signed-off-by: David Korczynski <david@adalogics.com>

### CODE REVIEW

**Comment by DavidKorczynski** [comment]
Compiling, running and extracting coverage of the tensorflow fuzztest fuzzers work:

tensorflow-fuzztest-coverage

Some refinement and cleanup is needed.

**Comment by DavidKorczynski** [comment]
CI is failing because of timeouts, however, in UBSAN we get past the point where the UBSAN FuzzTest fuzzers are build (showing the building works):
```
2022-11-16T23:47:38.9367669Z [32m[565 / 565][0m checking cached actions
2022-11-16T23:47:38.9442114Z 
2022-11-16T23:47:38.9442589Z [1A[K[32mINFO:[0m Build completed successfully, 565 total actions
2022-11-16T23:47:38.9442820Z 
2022-11-16T23:47:38.9443069Z [1A[K[32mINFO:[0m Build completed successfully, 565 total actions
2022-11-16T23:47:38.9453814Z [0m+ for fuzz_main_file in $FUZZ_TEST_BINARIES_OUT_PATHS
2022-11-16T23:47:38.9461793Z ++ bazel-out/k8-opt/bin/tensorflow/security/fuzzing/cc/status_fuzz --list_fuzz_tests
2022-11-16T23:47:38.9522466Z + FUZZ_TESTS=CC_FUZZING.FuzzTest
2022-11-16T23:47:38.9522995Z + cp bazel-out/k8-opt/bin/tensorflow/security/fuzzing/cc/status_fuzz /out/
2022-11-16T23:47:38.9704946Z ++ basename bazel-out/k8-opt/bin/tensorflow/security/fuzzing/cc/status_fuzz
2022-11-16T23:47:38.9721314Z + fuzz_basename=status_fuzz
2022-11-16T23:47:38.9721686Z + chmod -x /out/status_fuzz
2022-11-16T23:47:38.9739806Z + for fuzz_entrypoint in $FUZZ_TESTS
2022-11-16T23:47:38.9740158Z + TARGET_FUZZER=status_fuzz@CC_FUZZING.FuzzTest
2022-11-16T23:47:38.9740499Z + echo '#!/bin/sh
2022-11-16T23:47:38.9740767Z # LLVMFuzzerTestOneInput for fuzzer detection.
2022-11-16T23:47:38.9741036Z this_dir=$(dirname "$0")
2022-11-16T23:47:38.9741259Z chmod +x $this_dir/status_fuzz
2022-11-16T23:47:38.9741603Z status_fuzz --fuzz=CC_FUZZING.FuzzTest -- $@'
2022-11-16T23:47:38.9741900Z + chmod +x /out/status_fuzz@CC_FUZZING.FuzzTest
2022-11-16T23:47:38.9746754Z + for fuzz_main_file in $FUZZ_TEST_BINARIES_OUT_PATHS
2022-11-16T23:47:38.9750779Z ++ bazel-out/k8-opt/bin/tensorflow/security/fuzzing/cc/parseURI_fuzz --list_fuzz_tests
2022-11-16T23:47:38.9794784Z + FUZZ_TESTS=CC_FUZZING.FuzzTest
2022-11-16T23:47:38.9795308Z + cp bazel-out/k8-opt/bin/tensorflow/security/fuzzing/cc/parseURI_fuzz /out/
2022-11-16T23:47:38.9876186Z ++ basename bazel-out/k8-opt/bin/tensorflow/security/fuzzing/cc/parseURI_fuzz
2022-11-16T23:47:38.9884894Z + fuzz_basename=parseURI_fuzz
2022-11-16T23:47:38.9885275Z + chmod -x /out/parseURI_fuzz
2022-11-16T23:47:38.9894797Z + for fuzz_entrypoint in $FUZZ_TESTS
2022-11-16T23:47:38.9895099Z + TARGET_FUZZER=parseURI_fuzz@CC_FUZZING.FuzzTest
2022-11-16T23:47:38.9895451Z + echo '#!/bin/sh
```

**Comment by DavidKorczynski** [comment]
@oliverchang I switched this up to focus on fuzztest as it now has a potential fix for https://github.com/google/oss-fuzz/issues/8997 (was missing $ in front of this_dir in the wrapper script). I can do the Tensorflow aspects, which will be only in the Tensorflow project folder afterwards when https://github.com/tensorflow/tensorflow/pull/58646 is merged.

This is ready to go from my perspective.

File: infra/base-images/base-builder/compile_fuzztests.sh [hunk 1]
```diff
@@ -15,24 +15,34 @@
...
+fi
+
 # Trigger setup_configs rule of fuzztest as it generates the necessary
 # configuration file based on OSS-Fuzz environment variables.
-bazel run @com_google_fuzztest//bazel:setup_configs > fuzztest.bazelrc
+bazel run @com_google_fuzztest//bazel:setup_configs >> /etc/bazel.bazelrc
 
 # Bazel target names of the fuzz binaries.
-FUZZ_TEST_BINARIES=$(bazel query 'kind("cc_test", rdeps(..., @com_google_fuzztest//fuzztest:fuzztest_gtest_main))')
+#FUZZ_TEST_BINARIES=$(bazel query 'kind("cc_test", rdeps(..., @com_google_fuzztest//fuzztest:fuzztest_gtest_main))')
```
Comment (oliverchang) [inline infra/base-images/base-builder/compile_fuzztests.sh:38]: Can we just delete this commented line?

File: infra/base-images/base-builder/compile_fuzztests.sh [hunk 2]
```diff
@@ -52,13 +62,13 @@ for fuzz_main_file in $FUZZ_TEST_BINARIES_OUT_PATHS; do
...
 chmod +x \$this_dir/$fuzz_basename
-$fuzz_basename --fuzz=$fuzz_entrypoint -- \$@" > $OUT/$TARGET_FUZZER
+\$this_dir/$fuzz_basename --fuzz=$fuzz_entrypoint -- \$@" > $OUT/$TARGET_FUZZER
     chmod +x $OUT/$TARGET_FUZZER
   done
 done
 
 # synchronise coverage directory to bazel generated code.
-if [ "$SANITIZER" = "coverage" ]
+if [[ "$SANITIZER" = "coverage" && ${FUZZTEST_DO_SYNC:-"yes"} == "yes" ]]
```
Comment (oliverchang) [inline infra/base-images/base-builder/compile_fuzztests.sh:71]: Can you add more details on when `FUZZTEST_DO_SYNC` is needed in comments here?

File: infra/base-images/base-builder/compile_fuzztests.sh [hunk 3]
```diff
@@ -68,16 +78,20 @@ then
...
                     grep -v -x -F \
                     -e 'bazel-bin' \
-                    -e 'bazel-out' \
                     -e 'bazel-testlogs')"
   for link in $project_folders; do
    if [[ -d "${PWD}"/$link/external  ]]
    then
      rsync -avLk "${RSYNC_FILTER_ARGS[@]}" "${PWD}"/$link/external "${REMAP_PATH}"
    fi
+   if [[ -d "${PWD}"/$link/k8-opt  ]]
```
Comment (oliverchang) [inline infra/base-images/base-builder/compile_fuzztests.sh:91]: Likewise here -- can you add more context in comments for future readers of this code?

File: infra/base-images/base-builder/compile_fuzztests.sh [hunk 4]
```diff
@@ -15,24 +15,34 @@
 #
 ################################################################################
 
+set -x
+
+if [[ ${FUZZTEST_TARGET_FOLDER:-"unset"} == "unset" ]];
```
Comment (oliverchang) [inline infra/base-images/base-builder/compile_fuzztests.sh:20]: I think we need a bit more documentation on what this is needed for.

**Review by oliverchang — APPROVED** [review]
Thanks! This mostly LGTM. I think just needs some more clarification in terms of documentation.

**Comment by DavidKorczynski** [comment]
@oliverchang updated the docs, let me know what you think!

**Review by oliverchang — APPROVED** [review]

**Comment by oliverchang** [comment]
/gcbrun trial_build.py all --sanitizer address --fuzzing-engine libfuzzer

**Comment by oliverchang** [comment]
Thanks! Let's merge this on Monday to avoid any potential breakages over the weekend.

**Comment by oliverchang** [comment]
/gcbrun trial_build.py all --sanitizer address --fuzzing-engine libfuzzer

**Comment by oliverchang** [comment]
/gcbrun trial_build.py all --sanitizer address --fuzzing-engine libfuzzer

**Comment by oliverchang** [comment]
trial_build failures:


Step #1: bottleneck False
Step #1: numpy False
Step #1: skia False

Errors all look like spurious unrelated.  issues.

**Comment by DavidKorczynski** [comment]
@oliverchang am still not seeing results in the dashboard, is this and https://github.com/google/clusterfuzz/pull/2861 deployed?

**Comment by oliverchang** [comment]
Thanks for flagging @DavidKorczynski. We fixed a few more issues that came up in ClusterFuzz, and things should be moving along now. We should hopefully have things visible on the dashbboard in the next day or two.

**Pull request merged**